File: Depth-First Search/Binary tree/Structure variations/LET_no450_Delete Node in a BST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
        if not root:
            return None
        
        if root.val == key:
            if not root.right:
                nxt = root.left
                return nxt
            else:
                nxt = root.right
                while nxt.left:
                    nxt = nxt.left
                root.val, nxt.val = nxt.val, root.val
                root.right = self.deleteNode(root.right, key)
        elif key < root.val:
            root.left = self.deleteNode(root.left, key)
        else:
            root.right = self.deleteNode(root.right, key)
        
        return root

File: Depth-First Search/Binary tree/Structure variations/LET_no450_Delete Node in a BST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* deleteNode(TreeNode* root, int key) {
        if (!root)
            return nullptr;
        if (root->val == key) {
            if (!root->right) {
                TreeNode* nxt = root->left;
                delete root;
                return nxt;
            }
            else {
                TreeNode* nxt = root->right;
                while (nxt->left)
                    nxt = nxt->left;
                swap(nxt->val, root->val);
                root->right = deleteNode(root->right, key);
            }
            
        }
        else if (key < root->val)
            root->left = deleteNode(root->left, key);
        else
            root->right = deleteNode(root->right, key);
        
        return root;
    }
};